File: scripts/backup_dictionary_cache.js
/**
 * 备份 dictionary_cache 表数据到本地 JSON 文件
 *
 * 用途：
 * - 在执行批量更新前备份
 * - 定期备份防止数据丢失
 * - 迁移或恢复时使用
 *
 * 使用方法：
 *   node scripts/backup_dictionary_cache.js
 */

import { createClient } from '@supabase/supabase-js'
import fs from 'fs'
import path from 'path'

const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL
const supabaseKey = process.env.SUPABASE_SERVICE_ROLE_KEY

if (!supabaseUrl || !supabaseKey) {
  console.error('❌ 缺少环境变量: NEXT_PUBLIC_SUPABASE_URL 或 SUPABASE_SERVICE_ROLE_KEY')
  process.exit(1)
}

const supabase = createClient(supabaseUrl, supabaseKey)

// 备份目录
const BACKUP_DIR = path.join(process.cwd(), 'backups')
if (!fs.existsSync(BACKUP_DIR)) {
  fs.mkdirSync(BACKUP_DIR, { recursive: true })
}

async function backupTable(tableName) {
  console.log(`📦 开始备份表: ${tableName}`)

  const pageSize = 1000
  let allData = []
  let page = 0
  let hasMore = true

  while (hasMore) {
    const { data, error } = await supabase
      .from(tableName)
      .select('*')
      .range(page * pageSize, (page + 1) * pageSize - 1)

    if (error) {
      console.error(`❌ 查询失败 (第 ${page} 页):`, error.message)
      return
    }

    if (data && data.length > 0) {
      allData = allData.concat(data)
      console.log(`  ✓ 已获取 ${allData.length} 条记录...`)
      page++

      // 如果返回的数据少于 pageSize，说明已经是最后一页
      if (data.length < pageSize) {
        hasMore = false
      }
    } else {
      hasMore = false
    }
  }

  // 生成备份文件名（带时间戳）
  const timestamp = new Date().toISOString().replace(/[:.]/g, '-').split('T')[0]
  const filename = `${tableName}_${timestamp}.json`
  const filepath = path.join(BACKUP_DIR, filename)

  // 写入文件
  fs.writeFileSync(filepath, JSON.stringify(allData, null, 2), 'utf-8')

  console.log(`✅ 备份完成: ${filename}`)
  console.log(`   记录数: ${allData.length}`)
  console.log(`   文件大小: ${(fs.statSync(filepath).size / 1024 / 1024).toFixed(2)} MB`)

  return { filename, count: allData.length }
}

async function restoreTable(tableName, backupFile) {
  console.log(`📥 开始恢复表: ${tableName}`)

  // 读取备份文件
  const filepath = path.join(BACKUP_DIR, backupFile)
  if (!fs.existsSync(filepath)) {
    console.error(`❌ 备份文件不存在: ${backupFile}`)
    return
  }

  const data = JSON.parse(fs.readFileSync(filepath, 'utf-8'))
  console.log(`  读取到 ${data.length} 条记录`)

  // 清空表（谨慎操作！）
  console.log(`⚠️  即将清空表 ${tableName}，请确认...`)
  // const { error: deleteError } = await supabase.from(tableName).delete().neq('id', '00000000-0000-0000-0000-000000000000')

  // 插入数据
  const batchSize = 100
  for (let i = 0; i < data.length; i += batchSize) {
    const batch = data.slice(i, i + batchSize)
    const { error } = await supabase.from(tableName).insert(batch)

    if (error) {
      console.error(`❌ 插入失败 (第 ${i}-${i + batch.length} 条):`, error.message)
      return
    }

    console.log(`  ✓ 已插入 ${Math.min(i + batchSize, data.length)}/${data.length} 条记录`)
  }

  console.log(`✅ 恢复完成`)
}

// 主函数
async function main() {
  const command = process.argv[2] || 'backup'
  const table = process.argv[3] || 'dictionary_cache'

  if (command === 'backup') {
    await backupTable(table)
  } else if (command === 'restore') {
    const backupFile = process.argv[4]
    if (!backupFile) {
      console.error('❌ 请指定备份文件名')
      console.log('用法: node backup_dictionary_cache.js restore dictionary_cache backup_file.json')
      process.exit(1)
    }
    await restoreTable(table, backupFile)
  } else {
    console.log('用法:')
    console.log('  备份: node backup_dictionary_cache.js backup [table_name]')
    console.log('  恢复: node backup_dictionary_cache.js restore [table_name] [backup_file.json]')
  }
}

main()

File: scripts/backup_dictionary_cache.py
#!/usr/bin/env python3
"""
备份 dictionary_cache 表
导出结构和数据到 SQL 文件
"""

import os
import sys
import json
from pathlib import Path
from datetime import datetime
from typing import List, Dict
from supabase import create_client

# 加载环境变量
env_path = Path('/Users/a/dictation/.env.local')
with open(env_path) as f:
    for line in f:
        line = line.strip()
        if line and not line.startswith('#') and '=' in line:
            key, value = line.split('=', 1)
            os.environ[key.strip()] = value.strip()

# Supabase 配置
SUPABASE_URL = os.environ.get("NEXT_PUBLIC_SUPABASE_URL")
SUPABASE_KEY = os.environ.get("SUPABASE_SERVICE_ROLE_KEY")

if not SUPABASE_KEY:
    print("❌ 错误: 未找到 SUPABASE_SERVICE_ROLE_KEY")
    sys.exit(1)

print("🔗 连接 Supabase...")
client = create_client(SUPABASE_URL, SUPABASE_KEY)
print("✅ 连接成功\n")

# 备份目录
backup_dir = Path('/Users/a/dictation/backups')
backup_dir.mkdir(exist_ok=True)

# 生成备份文件名（带日期）
date_str = datetime.now().strftime('%Y%m%d')
backup_file = backup_dir / f'dictionary_cache_backup_{date_str}.sql'

print("=" * 70)
print("dictionary_cache 表备份")
print("=" * 70)
print(f"📁 备份文件: {backup_file}")
print()

# 获取所有数据
print("📖 正在获取数据...")
try:
    response = client.table('dictionary_cache').select('*').execute()

    if not response.data:
        print("⚠️  表为空或不存在")
        # 仍然创建备份文件（结构信息）
        with open(backup_file, 'w', encoding='utf-8') as f:
            f.write(f"-- dictionary_cache 表备份\n")
            f.write(f"-- 日期: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n")
            f.write(f"-- 来源: {SUPABASE_URL}\n")
            f.write(f"-- 状态: 表为空或不存在\n")
        print(f"✅ 空表备份已创建: {backup_file}")
        sys.exit(0)

    records = response.data
    print(f"✅ 获取到 {len(records)} 条记录")

except Exception as e:
    print(f"❌ 获取数据失败: {e}")
    sys.exit(1)

# 生成 SQL 备份文件
print(f"💾 正在生成 SQL 备份文件...")

with open(backup_file, 'w', encoding='utf-8') as f:
    # 写入头部信息
    f.write(f"-- dictionary_cache 表备份\n")
    f.write(f"-- 日期: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n")
    f.write(f"-- 来源: {SUPABASE_URL}\n")
    f.write(f"-- 记录数: {len(records)}\n")
    f.write(f"-- \n")
    f.write(f"-- 注意: 这是数据备份，不包含表结构定义\n")
    f.write(f"-- \n")
    f.write(f"-- 如需恢复，请使用 Supabase Dashboard 或 API\n")
    f.write(f"-- \n")
    f.write(f"BEGIN;\n\n")

    # 写入每条记录的 INSERT 语句
    for i, record in enumerate(records, 1):
        word = (record.get('word') or '').replace("'", "''")
        phonetic = (record.get('phonetic') or '').replace("'", "''")
        hit_count = record.get('hit_count', 0) or 0

        # 处理 definitions 字段（JSONB）
        definitions = record.get('definitions') or {}
        if definitions:
            definitions_json = json.dumps(definitions).replace("'", "''")
        else:
            definitions_json = '{}'

        example = (record.get('example') or '').replace("'", "''")
        audio_url_us = (record.get('audio_url_us') or '').replace("'", "''")
        audio_url_uk = (record.get('audio_url_uk') or '').replace("'", "''")

        # 生成 INSERT 语句
        f.write(f"-- 记录 {i}/{len(records)}\n")
        f.write(f"INSERT INTO dictionary_cache (word, phonetic, definitions, example, audio_url_us, audio_url_uk, hit_count) VALUES (\n")
        f.write(f"  '{word}',\n")
        f.write(f"  '{phonetic}',\n")
        f.write(f"  '{definitions_json}'::jsonb,\n")
        f.write(f"  '{example}',\n")
        f.write(f"  '{audio_url_us}',\n")
        f.write(f"  '{audio_url_uk}',\n")
        f.write(f"  {hit_count}\n")
        f.write(f");\n\n")

        # 每 100 条记录显示一次进度
        if i % 100 == 0:
            print(f"  处理中: {i}/{len(records)}")

    f.write("COMMIT;\n")
    f.write(f"\n-- 备份完成: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n")

# 检查文件大小
file_size = backup_file.stat().st_size
print(f"✅ 备份文件已生成: {backup_file}")
print(f"📊 文件大小: {file_size:,} bytes ({file_size/1024/1024:.2f} MB)")

# 验证文件内容
print(f"\n🔍 验证备份内容...")
with open(backup_file, 'r', encoding='utf-8') as f:
    lines = f.readlines()
    insert_count = sum(1 for line in lines if 'INSERT INTO' in line)

print(f"  - 总行数: {len(lines)}")
print(f"  - INSERT 语句: {insert_count}")

if insert_count == len(records):
    print(f"✅ 验证成功: 所有 {len(records)} 条记录都已备份")
else:
    print(f"⚠️  警告: INSERT 语句数量 ({insert_count}) 与记录数量 ({len(records)}) 不匹配")

print("\n" + "=" * 70)
print("🎉 备份完成！")
print("=" * 70)
